src/daft-logical-plan/src/optimization/rules/push_down_anti_semi_join.rs

```diff
@@ -0,0 +1,673 @@
+                        let mut requires_left = false;
+                        let mut requires_right = false;
+
+                        if let Some(on_expr) = on.inner() {
+                            on_expr
+                                .apply(|e| {
+                                    if let Expr::Column(Column::Resolved(
+                                        ResolvedColumn::JoinSide(field, JoinSide::Left),
+                                    )) = e.as_ref()
+                                    {
+                                        if child_left.schema().has_field(&field.name) {
+                                            requires_left = true;
+                                        }
+                                        if child_right.schema().has_field(&field.name) {
+                                            requires_right = true;
+                                        }
+                                    }
+
+                                    Ok(TreeNodeRecursion::Continue)
+                                })
+                                .unwrap();
+                        }
+
+                        match (requires_left, requires_right) {
+                            // requires only left side of child join, push down left
+                            (true, false) => {
+                                let new_child_left = Join::try_new(
+                                    child_left.clone(),
+                                    right.clone(),
+                                    on.clone(),
+                                    *join_type,
+                                    *join_strategy,
+                                )?
+                                .into();
+
+                                return Ok(Transformed::yes(Arc::new(
+                                    left.with_new_children(&[new_child_left, child_right.clone()]),
+                                )));
+                            }
+                            // requires only right side of child join, push down right
+                            (false, true) => {
+                                let new_child_right = Join::try_new(
+                                    child_right.clone(),
+                                    right.clone(),
+                                    on.clone(),
+                                    *join_type,
+                                    *join_strategy,
+                                )?
+                                .into();
+
+                                return Ok(Transformed::yes(Arc::new(
+                                    left.with_new_children(&[child_left.clone(), new_child_right]),
+                                )));
+                            }
+                            _ => {}
+                        }
+                    }
+                    // Anti/semi join can be trivially pushed down these ops
+                    LogicalPlan::Filter(Filter { input, .. })
```

Comment:
I think you can push down many more! I think it may make sense to add an method or macro to the logical plan for `can_push_down_predicate`. Since you should be able to push down anything that a filter can be pushed through. additional Examples include `repartition` and `concat`.

Reply:
I did consider implementing it for repartition and concat but chose not to.
- repartition - I am wary of reordering repartitions because they are usually manually added by the user, oftentimes to tune subsequent shuffles such as in joins.
- concat - we would have to push it down both sides of the concat, which means adding an additional join. The decision for whether to do that should probably be made in a cost-based rule instead.